Debug assertion `ASSERT(editingRange.location != notFound)` is frequently hit in Mail compose

rr-codes opened the pull request:
#### 88b600d040a8c6caf2c28647720de8087aad63de
<pre>
Debug assertion `ASSERT(editingRange.location != notFound)` is frequently hit in Mail compose
<a href="https://bugs.webkit.org/show_bug.cgi?id=276403">https://bugs.webkit.org/show_bug.cgi?id=276403</a>
<a href="https://rdar.apple.com/131427649">rdar://131427649</a>

Reviewed by Ryosuke Niwa.

The pattern of clients calling `selectedRangeWithCompletionHandler`, then passing the resulting range
to `attributedSubstringForProposedRange` is not uncommon and happens in several circumstances (for example,
AppKit does this when replying to an email in Mail, and WritingTools does this when showing the affordance).

Since it is perfectly valid for `selectedRangeWithCompletionHandler` to return `NSNotFound`, and also valid
for `attributedSubstringForProposedRange` to receive `NSNotFound`, and also results in valid behavior (when
it is notFound, the behavior is just to return the whole range of the editable content), this assertion can
and should be removed.

* Source/WebKit/Shared/EditingRange.cpp:
(WebKit::EditingRange::toRange):

Canonical link: <a href="https://commits.webkit.org/280817@main">https://commits.webkit.org/280817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cb7c1b8292ecad234c321c886a5e9c6303a0b27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57738 "Failed to checkout and rebase branch from PR 30629") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10214 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61360 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8183 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8371 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/5803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34762 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/49892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/27606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/31532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7187 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53484 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/7455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63042 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1652 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1658 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/49903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/54119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/1398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8604 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32895 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33981 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35065 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/Shared/EditingRange.cpp:
```diff
@@ -35,7 +35,6 @@ namespace WebKit {
 
 std::optional<WebCore::SimpleRange> EditingRange::toRange(WebCore::LocalFrame& frame, const EditingRange& editingRange, EditingRangeIsRelativeTo base)
 {
-    ASSERT(editingRange.location != notFound);
```
Can't we just return std::nullopt when it's notFound? What's the point of running the rest of the function?

rr-codes replied:
When it is notFound, the behavior is just to return the whole range of the editable content, which is different than returning null opt (which results in just nothing happening and the caller returning an empty string, which is not desired)

rr-codes replied:
(ie, the function still does something and returns a meaningful range)

rniwa replied:
In that case, we should explain that in the commit message.

rr-codes committed:
Debug assertion `ASSERT(editingRange.location != notFound)` is frequently hit in Mail compose
https://bugs.webkit.org/show_bug.cgi?id=276403
rdar://131427649

Reviewed by Ryosuke Niwa.

The pattern of clients calling `selectedRangeWithCompletionHandler`, then passing the resulting range
to `attributedSubstringForProposedRange` is not uncommon and happens in several circumstances (for example,
AppKit does this when replying to an email in Mail, and WritingTools does this when showing the affordance).

Since it is perfectly valid for `selectedRangeWithCompletionHandler` to return `NSNotFound`, and also valid
for `attributedSubstringForProposedRange` to receive `NSNotFound`, and also results in valid behavior (when
it is notFound, the behavior is just to return the whole range of the editable content), this assertion can
and should be removed.

* Source/WebKit/Shared/EditingRange.cpp:
(WebKit::EditingRange::toRange):

Canonical link: https://commits.webkit.org/280817@main
